### PR TITLE
fix(jmeter-service): Finish processes when '.finished' event is sent

### LINKED ISSUE #6778: Jmeter-service sends two sh.keptn.event.test.finished events when the endpoint is not reachable
Your issue may already be reported! Please search on the [issue tracker](https://github.com/keptn/keptn/issues) before creating a new one.

**Environment**

* Client OS (e.g., Linux, macOS, Windows): Linux
* Keptn Version (`keptn version`): 0.12.0
* Kubernetes Cloud Provider (e.g., GKE, AKS): GKE
* Kubernetes version (`kubectl version`): 1.21.6-gke.1500

**Affected Component**

* [ ] Approval
* [ ] Bridge
* [ ] CLI
* [ ] Datastore
* [ ] Docs
* [ ] Distributor / Event Broker
* [ ] Helm
* [ ] Istio
* [x] Jmeter
* [ ] Kubernetes Integration
* [ ] Openshift Integration
* [ ] REST API
* [ ] Webhook

**Describe the bug**

When triggering test for a deployment that is not reachable, jmeter-service sends the `.finished` event twice:
- the first one has result `fail` and message `Jmeter-service cannot reach URL fooservice: operation unsuccessful after 3 retry` 
- the second one has result `succeeded` and message `Tests for functional with status = true. Project: pchilafoobar4, Service: fooservice, Stage: dev, TestStrategy: functional, Context: 3bad94fb-5b84-41f2-930c-b5ed51c0c376` (this is probably because no jmeter files are present for the stage).

**To Reproduce**

1. Trigger a functional test for endpoints that are not reachable and without any jmeter files
2. wait for the test task to finish

**Expected behavior**

There should be only one `sh.keptn.event.test.finished` generated.

**Current behavior**

We get 2 `finished` events, one for the reachability test and one for the task completion

**Screenshots**

image

**Additional context**

First `finished event`

```
{
  "data": {
    "labels": {
      "buildId": "20220204.5",
      "runby": "Paolo Chila"
    },
    "message": "Jmeter-service cannot reach URL fooservice: operation unsuccessful after 3 retry",
    "project": "pchilafoobar4",
    "result": "fail",
    "service": "fooservice",
    "stage": "dev",
    "status": "succeeded",
    "test": {
      "end": "2022-02-04T09:03:33.703Z",
      "gitCommit": "",
      "start": "2022-02-04T09:03:18.195Z"
    }
  },
  "id": "5d021743-bfc5-4e4a-82b6-108f2718faf1",
  "source": "jmeter-service",
  "specversion": "1.0",
  "time": "2022-02-04T09:03:33.753Z",
  "type": "sh.keptn.event.test.finished",
  "shkeptncontext": "3bad94fb-5b84-41f2-930c-b5ed51c0c376",
  "shkeptnspecversion": "0.2.3",
  "triggeredid": "43ddf634-1524-48f6-82b5-630880a9dc54"
}
```

Second event:
```
{
  "data": {
    "labels": {
      "buildId": "20220204.5",
      "runby": "Paolo Chila"
    },
    "message": "Tests for functional with status = true. Project: pchilafoobar4, Service: fooservice, Stage: dev, TestStrategy: functional, Context: 3bad94fb-5b84-41f2-930c-b5ed51c0c376",
    "project": "pchilafoobar4",
    "result": "pass",
    "service": "fooservice",
    "stage": "dev",
    "status": "succeeded",
    "test": {
      "end": "2022-02-04T09:03:34.476Z",
      "gitCommit": "",
      "start": "2022-02-04T09:03:18.195Z"
    }
  },
  "id": "84a6ba77-8eea-4bec-b990-8df461a00c79",
  "source": "jmeter-service",
  "specversion": "1.0",
  "time": "2022-02-04T09:03:34.521Z",
  "type": "sh.keptn.event.test.finished",
  "shkeptncontext": "3bad94fb-5b84-41f2-930c-b5ed51c0c376",
  "shkeptnspecversion": "0.2.3",
  "triggeredid": "43ddf634-1524-48f6-82b5-630880a9dc54"
}
```

### PR DESCRIPTION
**Pull request by odubajDT**
Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

## This PR

- jmeter-service should finish processes and return from functions when '.finished' event is sent

### Related Issues

Fixes #6778 

Integration tests: https://github.com/keptn/keptn/actions/runs/1806186583

### CODE REVIEW

File: jmeter-service/testrunner.go [hunk 1]
```diff
@@ -149,7 +152,7 @@ func (tr *TestRunner) runHealthCheck(testInfo TestInfo, testStartedAt time.Time,
 		if err := tr.sendTestsFinishedEvent(testInfo, testStartedAt, msg, keptnv2.ResultFailed); err != nil {
 			logger.Errorf(errMsgSendFinishedEvent, err, testInfo)
 		}
-		return nil
+		return errors.New(msg)
```
Comment (bacherfl) [inline jmeter-service/testrunner.go:155]: I think we can keep `return nil` here, as a failed jmeter test is not necessarily an error in the execution of the `test` task

File: jmeter-service/testrunner.go [hunk 2]
```diff
@@ -158,14 +161,14 @@ func (tr *TestRunner) runHealthCheck(testInfo TestInfo, testStartedAt time.Time,
...
-		return nil
+		return err
 	}
 	if !res {
 		msg := fmt.Sprintf("Tests for %s with status = %s. %v", TestStrategy_HealthCheck, strconv.FormatBool(res), testInfo)
 		if err := tr.sendTestsFinishedEvent(testInfo, testStartedAt, msg, keptnv2.ResultFailed); err != nil {
 			logger.Errorf(errMsgSendFinishedEvent, err, testInfo)
 		}
-		return nil
+		return errors.New(msg)
```
Comment (bacherfl) [inline jmeter-service/testrunner.go:171]: I think we can keep `return nil` here, as a failed jmeter test is not necessarily an error in the execution of the `test` task

File: jmeter-service/testrunner.go (hunk 1 above)
Comment (odubajDT) [reply jmeter-service/testrunner.go:155]: When leaving nil here, we see that the finished event is sent in this place and then also in sendTestResult() function

Comment (bacherfl) [reply jmeter-service/testrunner.go:155]: ok, thx for clarifying :)

**Review by bacherfl — APPROVED** [review]

**Pull request merged**